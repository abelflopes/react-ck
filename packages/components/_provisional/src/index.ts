export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
