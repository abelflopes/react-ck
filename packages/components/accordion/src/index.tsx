import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
