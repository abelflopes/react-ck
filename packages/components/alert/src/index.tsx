import React from "react";
import classNames from "classnames";
import styles from "./styles/index.module.scss";
import { Text } from "@react-ck/text";

export interface AlertProps extends React.HTMLAttributes<HTMLDivElement> {
  heading?: string;
  /** Specifies the visual style of the alert  */
  skin?: "neutral" | "primary" | "negative" | "average" | "positive";
}

// TODO: add dismiss

/**
 * Alert is a short message that provides contextual feedback in a prominent way.
 * @param props - {@link AlertProps}
 * @returns a React element
 */

export const Alert = ({
  heading,
  skin = "neutral",
  children,
  className,
  ...otherProps
}: Readonly<AlertProps>): React.ReactElement => (
  <div {...otherProps} className={classNames(styles.root, styles[skin], className)}>
    {heading && (
      <Text type="h4" as="p" variation="bold" margin={false} className={styles.heading}>
        {heading}
      </Text>
    )}
    {children}
  </div>
);
