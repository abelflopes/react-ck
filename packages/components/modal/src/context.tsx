import React from "react";

interface HeaderProps extends Omit<React.HTMLAttributes<HTMLElement>, "children"> {
  heading?: string;
}

export interface ModalContextValue {
  header: HeaderProps | undefined;
  footer: React.HTMLAttributes<HTMLElement> | undefined;
}

export interface ModalContextProps {
  setValue: (value: Partial<ModalContextValue>) => void;
}

export const themeContextDefaults: ModalContextProps = {
  setValue: () => undefined,
};

export const ModalContext = React.createContext<ModalContextProps>(themeContextDefaults);
