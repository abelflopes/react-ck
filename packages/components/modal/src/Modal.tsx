import classNames from "classnames";
import styles from "./styles/index.module.scss";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import { ModalContext, type ModalContextProps, type ModalContextValue } from "./context";
import { Card } from "@react-ck/card";
import { Overlay } from "@react-ck/overlay";
import { Icon } from "@react-ck/icon";
import { Button } from "@react-ck/button";
import { Text } from "@react-ck/text";
import { Layer } from "@react-ck/layers";

/**
 * ModalProps interface represents the properties for the Modal component.
 * Extends React.HTMLAttributes<HTMLDivElement> to inherit HTMLDivElement attributes.
 */

interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Determines whether the modal is open or closed */
  open?: boolean;
  /** Determines if the modal can be dismissed by clicking outside or close button  */
  dismissable?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}

// TODO: add a11y https://react.dev/reference/react-dom/createPortal#rendering-a-modal-dialog-with-a-portal

/**
 * Modal is a full screen overlay that sits atop the page content.
 * Itâ€™s used to focus attention on an important task or message and requires user input to be dismissed.
 * @param props - {@link React.HTMLAttributes}
 * @returns a React element
 */

export const Modal = ({
  open = true,
  dismissable = true,
  children,
  className,
  onOpen,
  onClose,
  ...otherProps
}: Readonly<ModalProps>): React.ReactNode => {
  // Internal open state, allows component to be uncontrolled
  const [internalOpen, setInternalOpen] = useState(open);

  // State child compound components' props
  const [props, setProps] = useState<ModalContextValue>({
    header: undefined,
    footer: undefined,
  });

  // Define context properties from child compound components
  const setContextValue = useCallback<ModalContextProps["setValue"]>((value) => {
    setProps((v) => ({ ...v, ...value }));
  }, []);

  // Close callback
  const handleClose = useCallback(() => {
    if (!dismissable) return;
    setInternalOpen(false);
  }, [dismissable]);

  // Build context for child compound components
  const contextProps = useMemo<ModalContextProps>(
    () => ({
      setValue: setContextValue,
    }),
    [setContextValue],
  );

  // Invoke open / close handlers
  useEffect(() => {
    if (internalOpen) onOpen?.();
    else onClose?.();
  }, [internalOpen, onClose, onOpen]);

  // Synchronize internal state with external state
  useEffect(() => {
    setInternalOpen(open);
  }, [open]);

  return (
    internalOpen && (
      <Layer elevation="overlay">
        <div {...otherProps} className={classNames(styles.root, className)}>
          <Overlay
            className={classNames(dismissable && styles.clickable_overlay)}
            onClick={handleClose}
          />

          <Card className={styles.card}>
            <ModalContext.Provider value={contextProps}>
              {props.header ? (
                <header
                  {...props.header}
                  className={classNames(styles.header, props.header.className)}>
                  <Text variation="h3" as="p" margin={false}>
                    {props.header.heading}
                  </Text>

                  {dismissable ? (
                    <Button skin="ghost" icon={<Icon name="close" />} onClick={handleClose} />
                  ) : null}
                </header>
              ) : null}

              <main className={styles.content}>{children}</main>

              {props.footer ? (
                <footer
                  {...props.footer}
                  className={classNames(styles.footer, props.footer.className)}
                />
              ) : null}
            </ModalContext.Provider>
          </Card>
        </div>
      </Layer>
    )
  );
};
