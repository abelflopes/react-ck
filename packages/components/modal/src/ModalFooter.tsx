import { useContext, useEffect } from "react";
import { ModalContext, type ModalContextValue } from "./context";

export type ModalFooterProps = ModalContextValue["footer"];

export const ModalFooter = (props: NonNullable<ModalFooterProps>): undefined => {
  const context = useContext(ModalContext);

  useEffect(() => {
    context.setValue({ footer: props });

    return () => {
      context.setValue({ footer: undefined });
    };
  }, [context, props]);
};
