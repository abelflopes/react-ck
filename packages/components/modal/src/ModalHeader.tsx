import { useContext, useEffect } from "react";
import { ModalContext, type ModalContextValue } from "./context";

export const ModalHeader = (props: NonNullable<ModalContextValue["header"]>): undefined => {
  const context = useContext(ModalContext);

  useEffect(() => {
    context.setValue({ header: props });

    return () => {
      context.setValue({ header: undefined });
    };
  }, [context, props]);
};
