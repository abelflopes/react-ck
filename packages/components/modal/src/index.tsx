import classNames from "classnames";
import styles from "./styles/index.module.scss";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import { ModalContext, type ModalContextProps, type ModalContextValue } from "./context";
import { Card } from "@react-ck/card";
import { Overlay } from "@react-ck/overlay";
import { Icon } from "@react-ck/icon";
import { IconClose } from "@react-ck/icon/icons/IconClose";
import { Button } from "@react-ck/button";
import { Text } from "@react-ck/text";
import { Layer } from "@react-ck/layers";
import { ModalHeader } from "./ModalHeader";
import { ModalFooter } from "./ModalFooter";

/**
 * ModalProps interface represents the properties for the Modal component.
 * Extends React.HTMLAttributes<HTMLDivElement> to inherit HTMLDivElement attributes.
 */

interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Determines whether the modal is open or closed */
  open?: boolean;
  /** Determines if the modal can be dismissed by clicking outside or close button  */
  dismissible?: boolean;
  onDismiss?: () => void;
}

// TODO: add a11y https://react.dev/reference/react-dom/createPortal#rendering-a-modal-dialog-with-a-portal

let openModals = 0;

/**
 * Modal is a full screen overlay that sits atop the page content.
 * Itâ€™s used to focus attention on an important task or message and requires user input to be dismissed.
 * @param props - {@link React.HTMLAttributes}
 * @returns a React element
 */

const Modal = ({
  open = true,
  dismissible = true,
  children,
  className,
  onDismiss,
  ...otherProps
}: Readonly<ModalProps>): React.ReactNode => {
  // Internal open state, allows component to be uncontrolled
  const [internalOpen, setInternalOpen] = useState(open);

  // State child compound components' props
  const [props, setProps] = useState<ModalContextValue>({
    header: undefined,
    footer: undefined,
  });

  // Define context properties from child compound components
  const setContextValue = useCallback<ModalContextProps["setValue"]>((value) => {
    setProps((v) => ({ ...v, ...value }));
  }, []);

  // Close callback
  const handleClose = useCallback(() => {
    if (!dismissible) return;
    setInternalOpen(false);
    onDismiss?.();
  }, [dismissible, onDismiss]);

  // Build context for child compound components
  const contextProps = useMemo<ModalContextProps>(
    () => ({
      setValue: setContextValue,
    }),
    [setContextValue],
  );

  // Synchronize internal state with external state
  useEffect(() => {
    console.log("setInternalOpen", open);

    setInternalOpen(open);

    if (!open) onDismiss?.();
  }, [onDismiss, open]);

  useEffect(() => {
    console.log("internalOpen", internalOpen);
  }, [internalOpen]);

  // Lock scroll if there are open modals
  useEffect(() => {
    if (internalOpen) openModals += 1;
    else openModals -= 1;

    if (openModals > 1) document.body.classList.add(`${styles.lock_scroll}`);
    else document.body.classList.remove(`${styles.lock_scroll}`);
  }, [internalOpen]);

  return (
    internalOpen && (
      <Layer elevation="overlay">
        <div {...otherProps} className={classNames(styles.root, className)}>
          <Overlay
            className={classNames(dismissible && styles.clickable_overlay)}
            onClick={handleClose}
          />

          <Card className={styles.card}>
            <ModalContext.Provider value={contextProps}>
              {props.header ? (
                <header
                  {...props.header}
                  className={classNames(styles.header, props.header.className)}>
                  <Text variation="h3" as="p" margin="none">
                    {props.header.heading}
                  </Text>

                  {dismissible ? (
                    <Button
                      skin="ghost"
                      icon={
                        <Icon>
                          <IconClose />
                        </Icon>
                      }
                      onClick={handleClose}
                    />
                  ) : null}
                </header>
              ) : null}

              <main className={styles.content}>{children}</main>

              {props.footer ? (
                <footer
                  {...props.footer}
                  className={classNames(styles.footer, props.footer.className)}
                />
              ) : null}
            </ModalContext.Provider>
          </Card>
        </div>
      </Layer>
    )
  );
};

Modal.Header = ModalHeader;

Modal.Footer = ModalFooter;

export { Modal, type ModalProps };
