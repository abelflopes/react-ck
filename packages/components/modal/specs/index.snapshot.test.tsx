import React from "react";
import { Modal, ModalFooter, ModalHeader } from "../src/index";
import { LayersProvider } from "@react-ck/layers";
import { getActRender } from "@react-ck/jest-config";

describe("snapshot Modal", () => {
  it("renders correctly", async () => {
    const component = await getActRender(
      <LayersProvider>
        <Modal>
          <ModalHeader heading="Heading" />
          <ModalFooter>Footer</ModalFooter>
        </Modal>
      </LayersProvider>,
    );

    expect(component?.toJSON()).toMatchSnapshot();
  });
});
