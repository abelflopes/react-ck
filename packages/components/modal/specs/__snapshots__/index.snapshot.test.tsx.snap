// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Modal renders correctly 1`] = `
<div
  className="root"
>
  <div
    className="root dark clickable_overlay"
    onClick={[Function]}
  />
  <div
    className="root card bordered vertical spacing_m"
  >
    <div
      className="content"
    >
      <header
        className="header"
        heading="Heading"
      >
        <p
          className="root variation_h3 skin_default margin_none"
        >
          Heading
        </p>
        <button
          className="root ghost size-m icon-only"
          onClick={[Function]}
        >
          <svg
            className="root size_text skin_default"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3,3 L21,21 M3,21 L21,3"
              fill="none"
              strokeWidth="2"
            />
          </svg>
        </button>
      </header>
      <main
        className="content"
      />
      <footer
        className="footer"
      >
        Footer
      </footer>
    </div>
  </div>
</div>
`;
